pinot/inference/adlala.py

```diff
@@ -1 +1,215 @@
+        C: p := e^{-h * xi} p
+        D: p := p + sigma sqrt{h}R_n
+        E: xi := xi + h * epsilon [p^T p - N * tau]
+        O: p := c p + d R
+
+        
+        ```
+        NOTE: a B_(h/2) step is performed during initialization stage
+
+        """
+
+        # make sure closure is specified
+        assert closure is not None, 'Closure is needed in the training loop.'
+
+        # call closure
+        with torch.enable_grad():
+            loss = closure()
+
+        # loop through param groups with possible different parameter settings.
+        for group in self.param_groups:
+            for w in group['params']:
+                # NOTE: `w.grad == None` is different from `w.grad == 0.`
+                # as the later could be used to sample parameters without taking grads.
+                if w.grad is None: # skip network params not contributing to loss.
+                    continue
+
+                state = self.state[w]
+
+                # state initialization
+                if len(state) == 0:
+
+                    # initialize momentum
+                    state['p'] = torch.zeros_like(w)
+                    
+                    # initialize $\xi$
+                    state['xi'] = group['xi_init']
+
+                    # specify the $c$ and $d$ parameters for Langevin
+                    group['c'] = torch.exp(
+                            -group['h'] * group['gamma'])
+                    group['d'] = torch.sqrt(
+                            1. - torch.exp(-2. * group['h'] * group['gamma']
+                                )) * torch.sqrt(group['tau'])
+
+                    # do a B_(h/2) when during initialization
+                    # B_(h/2) step: `p := p - 0.5 * h * d_l`
+                    state['p'].add_(-0.5 * group['h'] * w.grad)
+                    
+                # A_(h/2) step: `w := theta + 0.5 * h * p`
+                w.add_(0.5 * group['h'] * state['p'])
+        
+        # C1 D1 E1 and O2 steps
+        for group in self.param_groups:
+            for w in group['params']:
+
+                if w.grad == None:
+                    continue
+
+                state = self.state[w]
+                if group['partition'].lower() == 'adla': # adaptive Langevin
+
+                    # C_(h/2) step: p := p * exp(-0.5 * h * \xi)
+                    state['p'].mul_(torch.exp(-0.5 * group['h'] * state['xi']))
+
+                    # D_(h/2) step: p := p + sqrt(0.5 * h) * \sigma * R
+                    state['p'].add_(
+                        torch.sqrt(0.5 * group['h']) *\
+                        group['sigma'] *\
+                        torch.randn(w.shape))
+                    
+                    # E_(h/2) step: \xi := \xi + 0.5 * h * \epsilon * (p^T p - N * \tao)
+                    state['xi'].add_(
+                        0.5 * group['h'] * group['epsilon'] *\
+                        (torch.sum(torch.pow(state['p'].flatten(), 2)) - state['p'].shape[0] * group['tau']))
```

Comment:
point of uncertainty: N here should be "number of parameters". is `state['p'].shape[0]` always equal to the number of parameters? may be safer to do `np.prod(state['p'].shape)` instead, in case `state['p']` is not flat

Reply:
`p` might not always be flat. in this case I'll switch to `torch.prod`

asked the authors here for further clarification

https://github.com/TiffanyVlaar/ThermodynamicParameterizationOfNNs/issues/1

Reply:
fixed in 7c7e09e